src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +39,489 @@ static int16_t axisval(int16_t x)
+            got_event = true;
+            break;
+        }
 
-        joystate[joy].instance = instance;
+        default:
+            // Ignore all other events.
+            break;
+        }
     }
 
-    joystate[joy].index = joy;
-    joystate[joy].is_gc = is_gc;
-}
-
-void wxSDLJoy::RemapControllers()
-{
-    for (auto&& joy : joystate) {
-        auto& state = joy.second;
+    bool do_poll = false;
+    wxLongLong now = wxGetUTCTimeMillis();
 
-        DisconnectController(state);
-        ConnectController(joy.first);
+    if (got_event) {
+        last_poll_ = now;
+    } else if (now - last_poll_ > kPollTimeInterval) {
+        do_poll = true;
+        last_poll_ = now;
     }
-}
 
-void wxSDLJoy::DisconnectController(wxSDLJoyState& state)
-{
-    if (auto& dev = state.dev) {
-        if (state.is_gc) {
-            SDL_GameControllerClose(dev);
-        }
-        else {
-            SDL_JoystickClose(dev);
+    if (do_poll) {
+        for (auto&& joy_state : joystick_states_) {
+            joy_state.second->Poll();
         }
-
-        dev = nullptr;
     }
-
-    instance_map.erase(state.instance);
 }
 
-void wxSDLJoy::Add(int8_t joy_n)
-{
-    if (joy_n < 0) {
-        for (uint8_t joy : range(0, SDL_NumJoysticks()))
-            ConnectController(joy);
-
-        add_all = true;
+void wxSDLJoy::DisconnectController(SDL_JoystickID joystick_index) {
+    assert(!add_all_joysticks_);
 
+    auto iter = joystick_states_.find(joystick_index);
+    if (iter == joystick_states_.end()) {
+        // Nothing to do if the joystick was not active.
         return;
     }
 
-    ConnectController(joy_n);
+    // Update needed SDL indexes.
+    int sdl_index = iter->second->player_index() - 1;
+    needed_sdl_indexes_.insert(sdl_index);
+    joystick_states_.erase(iter);
 }
 
-void wxSDLJoy::Remove(int8_t joy_n)
-{
-    add_all = false;
-
-    if (joy_n < 0) {
-        for (auto&& joy : joystate)
-            DisconnectController(joy.second);
-
-        joystate.clear();
-
-        return;
+void wxSDLJoy::RemapControllers() {
+    if (!add_all_joysticks_) {
+        // Save the connected SDL indexes first.
+        for (const auto& joy_state : joystick_states_) {
+            needed_sdl_indexes_.insert(joy_state.second->player_index() - 1);
+        }
     }
 
-    DisconnectController(joystate[joy_n]);
-    joystate.erase(joy_n);
-}
-
-void wxSDLJoy::SetRumble(bool do_rumble)
-{
-    rumbling = do_rumble;
-
-#if SDL_VERSION_ATLEAST(2, 0, 9)
-    // Do rumble only on device 0, and only if it's a GameController.
-    auto dev = joystate[0].dev;
-    if (dev && SDL_IsGameController(0)) {
-        if (rumbling) {
-            SDL_GameControllerRumble(dev, 0xFFFF, 0xFFFF, 300);
-            if (!IsRunning())
-                Start(150);
-        }
-        else {
-            SDL_GameControllerRumble(dev, 0, 0, 0);
-            Stop();
+    joystick_states_.clear();
+    for (int i = 0; i < SDL_NumJoysticks(); i++) {
```

Comment:
This should iterate over wanted SDL indices in the normal case and all available joysticks in the add-all case.

Reply:
As the code is currently, this is incorrect behavior, since on attach of a controller all available joysticks will be attached, but in the general case we only want the indexes requested from config.

Reply:
You are correct, I rewrote this method.